### PR TITLE
polygamma improvements (evaluation, real and positivity determination)

### LINKED ISSUE #12569: problem with polygamma or im
N(polygamma(0,I))
yields a correct result 
0.094650320622477 + 2.07667404746858*I
but
im(polygamma(0,I))
yields zero.

### LINKED ISSUE #17350: Should polygamma(a, b) raise NotImplementedError for noninteger (numeric) a?
mpmath's implementation of polygamma is valid only for integer first arguments. SymPy's implementation does not care, but this leads to many issues if you actually try to do things with such objects:

```
>>> (I*polygamma(I, pi)).as_real_imag()
Traceback (most recent call last):
  File "/home/eward/se2/sympy/core/assumptions.py", line 262, in getit
    return self._assumptions[fact]
KeyError: 'zero'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/eward/se2/sympy/core/assumptions.py", line 262, in getit
    return self._assumptions[fact]
KeyError: 'finite'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/eward/se2/sympy/core/assumptions.py", line 262, in getit
    return self._assumptions[fact]
KeyError: 'extended_positive'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/eward/se2/sympy/core/evalf.py", line 1308, in evalf
    rf = evalf_table[x.func]
KeyError: polygamma

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/eward/se2/sympy/core/mul.py", line 798, in as_real_imag
    if i.is_zero:
  File "/home/eward/se2/sympy/core/assumptions.py", line 266, in getit
    return _ask(fact, self)
  File "/home/eward/se2/sympy/core/assumptions.py", line 321, in _ask
    _ask(pk, obj)
  File "/home/eward/se2/sympy/core/assumptions.py", line 309, in _ask
    a = evaluate(obj)
  File "/home/eward/se2/sympy/core/expr.py", line 864, in _eval_is_negative
    finite = self.is_finite
  File "/home/eward/se2/sympy/core/assumptions.py", line 266, in getit
    return _ask(fact, self)
  File "/home/eward/se2/sympy/core/assumptions.py", line 321, in _ask
    _ask(pk, obj)
  File "/home/eward/se2/sympy/core/assumptions.py", line 309, in _ask
    a = evaluate(obj)
  File "/home/eward/se2/sympy/core/expr.py", line 857, in _eval_is_positive
    extended_positive = self.is_extended_positive
  File "/home/eward/se2/sympy/core/assumptions.py", line 266, in getit
    return _ask(fact, self)
  File "/home/eward/se2/sympy/core/assumptions.py", line 309, in _ask
    a = evaluate(obj)
  File "/home/eward/se2/sympy/core/expr.py", line 882, in _eval_is_extended_positive
    n2 = self._eval_evalf(2)
  File "/home/eward/se2/sympy/core/function.py", line 565, in _eval_evalf
    args = [arg._to_mpmath(prec + 5) for arg in self.args]
  File "/home/eward/se2/sympy/core/function.py", line 565, in <listcomp>
    args = [arg._to_mpmath(prec + 5) for arg in self.args]
  File "/home/eward/se2/sympy/core/evalf.py", line 1489, in _to_mpmath
    re, im, _, _ = evalf(self, prec, {})
  File "/home/eward/se2/sympy/core/evalf.py", line 1314, in evalf
    xe = x._eval_evalf(prec)
  File "/home/eward/se2/sympy/core/function.py", line 590, in _eval_evalf
    v = func(*args)
  File "/usr/local/lib/python3.6/dist-packages/mpmath/ctx_mp.py", line 265, in psi
    m = int(m)
TypeError: int() argument must be a string, a bytes-like object or a number, not 'mpc'
>>> (tanh(polygamma(I, 1))).rewrite(exp)
...
    m = int(m)
TypeError: int() argument must be a string, a bytes-like object or a number, not 'mpc'
>>> (I / polygamma(I, 4)).rewrite(exp)
...
    raise TypeError("Invalid comparison of complex %s" % me)
TypeError: Invalid comparison of complex I
```

### PR DESCRIPTION
**Pull request by ethankward**
Closes #17350. Fixes #12569.

#### Brief description of what is fixed or changed

- Add `_eval_evalf` to prevent evaluation when the first argument is not a nonnegative integer.
- Remove `_eval_is_real` since the logic was wrong.
- Update `_eval_is_positive` and `_eval_is_negative` to avoid doing comparisons with `>`.

#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/functions/special/gamma_functions.py [hunk 1]
```diff
@@ -583,6 +583,12 @@ class polygamma(Function):
     .. [4] http://functions.wolfram.com/GammaBetaErf/PolyGamma2/
     """
 
+    def _eval_evalf(self, prec):
+        n = self.args[0]
+        # the mpmath polygamma implementation valid only for nonnegative integers
+        if n.is_real and n.is_number:
```
Comment (jksuom) [inline sympy/functions/special/gamma_functions.py]: How about changing the order? I think that `n.is_number` will generally take less time than `n.is_real` (though I may be wrong).

Comment (oscargus) [reply sympy/functions/special/gamma_functions.py]: It should be correct that `n.is_number` is faster as it only has to check if there are any free symbols and not worry about the assumptions on the symbols.

**Comment by oscargus** [comment]
Looks good!

Does it make sense to be able to instantiate expressions like `polygamma(I, 2)`? Seems to have limited use...

Also, `polygamma(I, 2).rewrite(zeta)` raises an exception.

Not really related to the PR as such, but since there are tests with imaginary derivative order it may be that people later may hesitate to remove them...

**Comment by ethankward** [comment]
What would be the alternative instead of instantiating them? Raising a `NotImplementedError`? I didn't get any response to that in #17350.

**Comment by oscargus** [comment]
Good question! I've seen that there are indeed extensions to the complex plane, so should be fine. Especially now when you kindly fixed the zeta rewrite. Thanks!

**Pull request merged**